libs/framework/include/celix/Trackers.h

```diff
@@ -0,0 +1,783 @@
+#include "celix/Utils.h"
+#include "celix/Bundle.h"
+#include "celix/Constants.h"
+#include "celix/Filter.h"
+#include "celix_bundle_context.h"
+
+namespace celix {
+
+
+    /**
+     * @brief The tracker state.
+     */
+    enum class TrackerState {
+        OPENING,
+        OPEN,
+        CLOSING,
+        CLOSED
+    };
+
+    /**
+     * @brief The AbstractTracker class is the base of all C++ Celix trackers.
+     *
+     * It defines how trackers are closed and manages the tracker state.
+     *
+     * This class can be used to create a vector of different (shared ptr)
+     * trackers (i.e. ServiceTracker, BundleTracker, MetaTracker).
+     *
+     * AbstractTracker <-----------------------------------
+     *  ^                               |                |
+     *  |                               |                |
+     *  GenericServiceTracker      BundleTracker    MetaTracker
+     *  ^
+     *  |
+     *  ServiceTracker<I>
+     *
+     * @note Thread safe.
+     */
+    class AbstractTracker {
+    public:
+        explicit AbstractTracker(std::shared_ptr<celix_bundle_context_t> _cCtx) :
+            cCtx{std::move(_cCtx)} {}
+
+        virtual ~AbstractTracker() noexcept = default;
+
+        /**
+         * @brief Check if the tracker is open (state == OPEN)
+         */
+        bool isOpen() const {
+            std::lock_guard<std::mutex> lck{mutex};
+            return state == TrackerState::OPEN;
+        }
+
+        /**
+         * @brief Get the current state of the tracker.
+         */
+        TrackerState getState() const {
+            std::lock_guard<std::mutex> lck{mutex};
+            return state;
+        }
+
+        /**
+         * @brief Close the tracker (of the state is not CLOSED or CLOSING).
+         *
+         * This will be done sync so then the close() method return the
+         * tracker is closed and all the needed callbacks have been called.
+         */
+        void close() {
+            long localTrkId = -1;
+            {
+                std::lock_guard<std::mutex> lck{mutex};
+                if (state == TrackerState::OPEN || state == TrackerState::OPENING) {
+                    //not yet closed
+                    state = TrackerState::CLOSING;
+                    localTrkId = trkId;
+                    trkId = -1;
+                }
+            }
+            if (localTrkId >= 0) {
+                celix_bundleContext_stopTracker(cCtx.get(), localTrkId);
+                {
+                    std::lock_guard<std::mutex> lck{mutex};
+                    state = TrackerState::CLOSED;
+                }
+            }
+        }
+
+        /**
+         * @brief Open the tracker (if the state is not OPEN or OPENING).
+         *
+         * This is done async, meaning that the actual opening of the tracker will be
+         * done a a Celix event processed on the Celix event thread.
+         *
+         * @throws celix::Exception
+         */
+        virtual void open() {}
```

Comment:
Any particular reason this is not a pure virtual function?

Reply:
Nope, should be a pure virtual. updated.